src/envs/pddl_env.py

```diff
@@ -0,0 +1,442 @@
+            act_arr[1:len(obj_idxs) + 1] = obj_idxs
+            return Action(act_arr)
+
+        def initiable(s: State, m: Dict, o: Sequence[Object],
+                      p: Array) -> bool:
+            del m, p  # unused
+            ground_atoms = _state_to_ground_atoms(s)
+            ground_op = op.ground(tuple(o))
+            return ground_op.preconditions.issubset(ground_atoms)
+
+        return utils.SingletonParameterizedOption(name,
+                                                  policy,
+                                                  types,
+                                                  initiable=initiable)
+
+
+class _FixedTasksPDDLEnv(_PDDLEnv):
+    """An env where tasks are induced by static PDDL problem files."""
+
+    @property
+    @abc.abstractmethod
+    def _pddl_problem_asset_dir(self) -> str:
+        """The name of the problem file directory in assets/pddl."""
+        raise NotImplementedError("Override me!")
+
+    @property
+    @abc.abstractmethod
+    def _train_problem_indices(self) -> List[int]:
+        """The problem file names should be task{idx}.pddl."""
+        raise NotImplementedError("Override me!")
+
+    @property
+    @abc.abstractmethod
+    def _test_problem_indices(self) -> List[int]:
+        """The problem file names should be task{idx}.pddl."""
+        raise NotImplementedError("Override me!")
+
+    @property
+    def _pddl_train_problem_generator(self) -> PDDLProblemGenerator:
+        assert len(self._train_problem_indices) >= CFG.num_train_tasks
+        return _file_problem_generator(self._pddl_problem_asset_dir,
+                                       self._train_problem_indices)
+
+    @property
+    def _pddl_test_problem_generator(self) -> PDDLProblemGenerator:
+        assert len(self._test_problem_indices) >= CFG.num_test_tasks
+        return _file_problem_generator(self._pddl_problem_asset_dir,
+                                       self._test_problem_indices)
+
+
+###############################################################################
+#                              Concrete Envs                                  #
+###############################################################################
+
+
+class _BlocksPDDLEnv(_PDDLEnv):
+    """The IPC 4-operator blocks world domain."""
+
+    @property
+    def _domain_str(self) -> str:
+        path = utils.get_env_asset_path("pddl/blocks/domain.pddl")
+        with open(path, encoding="utf-8") as f:
+            domain_str = f.read()
+        return domain_str
+
+
+class FixedTasksBlocksPDDLEnv(_BlocksPDDLEnv, _FixedTasksPDDLEnv):
+    """The IPC 4-operator blocks world domain with a fixed set of tasks."""
+
+    @classmethod
+    def get_name(cls) -> str:
+        return "pddl_blocks_fixed_tasks"
+
+    @property
+    def _pddl_problem_asset_dir(self) -> str:
+        return "blocks"
+
+    @property
+    def _train_problem_indices(self) -> List[int]:
+        return list(range(1, 6))
```

Comment:
instead of magic numbers 6 and 11, let's grab the number of problems in the folder and use half for train, half for test?

Reply:
I prefer this way because I want it to be easy to experiment with different train and test problem sets. Manipulating the files themselves is a pain.

Reply:
what do you mean by manipulating the files? wouldn't it just be `len(os.listdir(path to blocks folder))`?

Reply:
I mean in the course of doing research, if I want to experiment with a certain subset of all the PDDL files I have, I don't want to have to change the files in my directory manually, I would rather change the numbers here